File: LMS_meeting/LMS/screenshot_analyzer.py
import os
import base64

try:
    import cv2
    import numpy as np
    HAS_OPENCV = True
except ImportError:
    HAS_OPENCV = False

try:
    from PIL import Image, ImageChops
    HAS_PIL = True
except ImportError:
    HAS_PIL = False

# Configurable thresholds & weights
Y_THRESHOLD = 30.0
R_THRESHOLD = 60.0
SS_WEIGHT = 0.3
POPUP_THRESHOLD = 70.0

def frame_diff_score(frame_a, frame_b) -> float:
    """
    Computes absolute pixel difference between frame_a and frame_b (grayscale 160x90).
    Returns a normalized change score in range 0-100.
    """
    if HAS_OPENCV:
        # Resize both frames to 160x90 (lightweight, fast)
        a = cv2.resize(frame_a, (160, 90))
        b = cv2.resize(frame_b, (160, 90))
        
        # Convert to grayscale
        a_gray = cv2.cvtColor(a, cv2.COLOR_BGR2GRAY) if len(a.shape) == 3 else a
        b_gray = cv2.cvtColor(b, cv2.COLOR_BGR2GRAY) if len(b.shape) == 3 else b
        
        # Compute absolute pixel difference
        diff = cv2.absdiff(a_gray, b_gray)
        mean_diff = float(np.mean(diff))
    elif HAS_PIL:
        if isinstance(frame_a, np.ndarray):
            img_a = Image.fromarray(frame_a).resize((160, 90)).convert('L')
            img_b = Image.fromarray(frame_b).resize((160, 90)).convert('L')
        else:
            img_a = frame_a.resize((160, 90)).convert('L')
            img_b = frame_b.resize((160, 90)).convert('L')
        
        diff = ImageChops.difference(img_a, img_b)
        stat = ImageChops.stat.Stat(diff)
        mean_diff = stat.mean[0]
    else:
        # Pure Python list/matrix calculation fallback
        a_flat = list(frame_a) if isinstance(frame_a, (list, tuple)) else frame_a.flatten()
        b_flat = list(frame_b) if isinstance(frame_b, (list, tuple)) else frame_b.flatten()
        diffs = [abs(int(p1) - int(p2)) for p1, p2 in zip(a_flat, b_flat)]
        mean_diff = sum(diffs) / max(len(diffs), 1)

    score = (mean_diff / 255.0) * 100.0
    return round(score, 2)

def compress_frame_480p(frame, quality: int = 50) -> str:
    """
    Compresses nearest screenshot frame to 480p resolution (854x480 JPEG quality 50%),
    resulting in a tiny payload (< 100KB, well under 1MB) for LMS storage.
    """
    if HAS_OPENCV and isinstance(frame, np.ndarray):
        # Resize to 480p (854x480)
        resized = cv2.resize(frame, (854, 480))
        encode_param = [int(cv2.IMWRITE_JPEG_QUALITY), quality]
        _, buffer = cv2.imencode('.jpg', resized, encode_param)
        jpg_as_text = base64.b64encode(buffer).decode('utf-8')
        return f"data:image/jpeg;base64,{jpg_as_text}"
    elif HAS_PIL:
        import io
        if isinstance(frame, np.ndarray):
            img = Image.fromarray(frame).resize((854, 480))
        else:
            img = frame.resize((854, 480))
        buf = io.BytesIO()
        img.save(buf, format='JPEG', quality=quality)
        jpg_as_text = base64.b64encode(buf.getvalue()).decode('utf-8')
        return f"data:image/jpeg;base64,{jpg_as_text}"
    else:
        return "data:image/jpeg;base64,compressed_480p_dummy_frame_data"

def process_screenshot_slot(recent_frame, 
                           previous_frame, 
                           behavioral_score: float = 0.0):
    """
    Processes 5s slot screenshot frame and returns audit decision and formatted score code (y<score> / r<score>).
    
    Returns dict with:
      - output_code: 'y<score>' or 'r<score>' (or None if below Y threshold)
      - event_type: 'SCREENSHOT_CHANGE' or 'SCREENSHOT_ANOMALY'
      - score: float (diff score)
      - combined_risk: float
      - compressed_evidence: str (JPEG base64 if >= Y_THRESHOLD, else None)
      - popup_triggered: bool
      - updated_previous_frame: np.ndarray
    """
    if previous_frame is None:
        return {
            "output_code": None,
            "score": 0.0,
            "combined_risk": behavioral_score,
            "event_type": None,
            "compressed_evidence": None,
            "popup_triggered": False,
            "updated_previous_frame": recent_frame
        }
        
    score = frame_diff_score(recent_frame, previous_frame)
    combined_risk = round(behavioral_score + (score * SS_WEIGHT), 2)
    
    if score < Y_THRESHOLD:
        # Negligible change — replace buffer, continue
        return {
            "output_code": None,
            "score": score,
            "combined_risk": combined_risk,
            "event_type": None,
            "compressed_evidence": None,
            "popup_triggered": False,
            "updated_previous_frame": recent_frame
        }
        
    elif Y_THRESHOLD <= score < R_THRESHOLD:
        # Yellow threshold: telemetry log ONLY — do NOT attach screenshot or fire popup
        output_code = f"y{score}"
        return {
            "output_code": output_code,
            "score": score,
            "combined_risk": combined_risk,
            "event_type": "SCREENSHOT_CHANGE",
            "compressed_evidence": None, # No screenshot attached for y<score>
            "zip_evidence": None,
            "popup_triggered": False,
            "updated_previous_frame": recent_frame
        }
        
    else:  # score >= R_THRESHOLD
        # Red threshold: grab nearest screenshot, compress to 480p resolution (<100KB), log event, and trigger popup
        output_code = f"r{score}"
        evidence_img_480p = compress_frame_480p(recent_frame, quality=50)
        
        return {
            "output_code": output_code,
            "score": score,
            "combined_risk": combined_risk,
            "event_type": "SCREENSHOT_ANOMALY",
            "compressed_evidence": evidence_img_480p, # 480p compressed image (<100KB, well under 1MB)
            "popup_triggered": True, # Trigger popup signal for r<score>
            "updated_previous_frame": recent_frame
        }

if __name__ == "__main__":
    print("--- Running Screenshot Frame Diff Anomaly Detection Test ---")
    
    if HAS_OPENCV:
        frame_a = np.zeros((720, 1280, 3), dtype=np.uint8)
        frame_b = np.copy(frame_a)
        cv2.rectangle(frame_b, (100, 100), (500, 500), (200, 200, 200), -1)
        frame_c = np.full((720, 1280, 3), 255, dtype=np.uint8)
    else:
        # Fallback 100px pixel arrays for testing
        frame_a = [0] * 100
        # Medium change (yellow)
        frame_b = [0] * 50 + [200] * 50
        # High change (red)
        frame_c = [255] * 100

    res_yellow = process_screenshot_slot(frame_b, frame_a, behavioral_score=15.0)
    print(f"[Yellow Test] Output Code: {res_yellow['output_code']}, Score: {res_yellow['score']}, Combined Risk: {res_yellow['combined_risk']}")
    
    res_red = process_screenshot_slot(frame_c, frame_a, behavioral_score=50.0)
    print(f"[Red Test] Output Code: {res_red['output_code']}, Score: {res_red['score']}, Combined Risk: {res_red['combined_risk']}, Popup: {res_red['popup_triggered']}")

File: LMS_meeting/LMS/frontend/src/utils/screenshotAnalyzer.js
/**
 * Screenshot Frame Difference Anomaly Analyzer (JS / Canvas implementation)
 * Operates on 3-second screenshot intervals.
 * Evaluates frame difference (160x90 grayscale), assigns y<score> or r<score>,
 * compresses evidence images, and sends to LMS audit logs.
 */

export const Y_THRESHOLD = 30.0;
export const R_THRESHOLD = 60.0;
export const SS_WEIGHT = 0.3;
export const POPUP_THRESHOLD = 70.0;

let previousFrameData = null;

/**
 * Resizes source image/canvas to 160x90 grayscale pixel array
 */
function getGrayscalePixels(sourceCanvas) {
  const offscreen = document.createElement('canvas');
  offscreen.width = 160;
  offscreen.height = 90;
  const ctx = offscreen.getContext('2d');
  
  // Draw scaled down
  ctx.drawImage(sourceCanvas, 0, 0, 160, 90);
  const imgData = ctx.getImageData(0, 0, 160, 90).data;
  
  // Convert RGBA to Grayscale array (length 160 * 90)
  const grayPixels = new Uint8Array(160 * 90);
  for (let i = 0, j = 0; i < imgData.length; i += 4, j++) {
    // Luminance formula: Y = 0.299R + 0.587G + 0.114B
    grayPixels[j] = Math.round(0.299 * imgData[i] + 0.587 * imgData[i + 1] + 0.114 * imgData[i + 2]);
  }
  
  return grayPixels;
}

/**
 * Calculates absolute pixel difference mean normalized to 0-100 score
 */
export function computeDiffScore(grayPixelsA, grayPixelsB) {
  if (!grayPixelsA || !grayPixelsB || grayPixelsA.length !== grayPixelsB.length) return 0;
  
  let totalDiff = 0;
  const len = grayPixelsA.length;
  for (let i = 0; i < len; i++) {
    totalDiff += Math.abs(grayPixelsA[i] - grayPixelsB[i]);
  }
  
  const meanDiff = totalDiff / len; // 0 - 255 range
  const score = (meanDiff / 255.0) * 100.0;
  return Number(score.toFixed(2));
}

/**
 * Compresses current canvas to 480p JPEG data URL (854x480, quality 50%)
 * Payload is tiny (< 100KB, well under 1MB) for LMS storage.
 */
export function compressCanvasFrame480p(canvas, quality = 0.5) {
  try {
    const offscreen = document.createElement('canvas');
    offscreen.width = 854;
    offscreen.height = 480;
    const ctx = offscreen.getContext('2d');
    ctx.drawImage(canvas, 0, 0, 854, 480);
    return offscreen.toDataURL('image/jpeg', quality);
  } catch (err) {
    console.error('Error compressing 480p frame:', err);
    return null;
  }
}

/**
 * Processes 3s slot screenshot frame
 * @param {HTMLCanvasElement} currentCanvas 
 * @param {number} behavioralCurrentScore 
 * @returns {object|null} Result with outputCode ('y<score>' | 'r<score>'), evidence frame, combinedRisk, eventType
 */
export function analyze3sSlotFrame(currentCanvas, behavioralCurrentScore = 0) {
  const currentGray = getGrayscalePixels(currentCanvas);
  
  if (!previousFrameData) {
    previousFrameData = currentGray;
    return {
      outputCode: null,
      score: 0,
      combinedRisk: behavioralCurrentScore,
      eventType: null,
      evidenceFrame: null,
      popupTriggered: false,
    };
  }
  
  const score = computeDiffScore(currentGray, previousFrameData);
  const combinedRisk = Number((behavioralCurrentScore + (score * SS_WEIGHT)).toFixed(2));
  previousFrameData = currentGray;
  
  if (score < Y_THRESHOLD) {
    // Negligible change (< 30): replace buffer, continue
    return {
      outputCode: null,
      score,
      combinedRisk,
      eventType: null,
      evidenceFrame: null,
      popupTriggered: false,
    };
  }

  if (score >= Y_THRESHOLD && score < R_THRESHOLD) {
    // Yellow: noteworthy change (30-60) — Telemetry log ONLY, no screenshot, no popup
    const outputCode = `y${score}`;
    return {
      outputCode,
      score,
      combinedRisk,
      eventType: 'SCREENSHOT_CHANGE',
      evidenceFrame: null, // NO screenshot attached for y<score>
      popupTriggered: false,
    };
  } else {
    // Red: large sudden visual change (>= 60) — Grab nearest 480p compressed screenshot & trigger popup
    const outputCode = `r${score}`;
    const compressed480p = compressCanvasFrame480p(currentCanvas, 0.5);
    
    return {
      outputCode,
      score,
      combinedRisk,
      eventType: 'SCREENSHOT_ANOMALY',
      evidenceFrame: compressed480p, // 480p compressed screenshot (< 100KB)
      popupTriggered: true, // Trigger popup signal for r<score>
    };
  }
}

export function resetFrameBuffer() {
  previousFrameData = null;
}

export const analyze5sSlotFrame = analyze3sSlotFrame;
